075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,435 @@
+Another option could be leaving both the ways living together.
+If this proposal is considered valid, doing such a change on the `KafkaRebalance` custom resource would help the proposal itself because the operator should not copy over the entire configuration from the "template" resource to the actual rebalance resource but just set the configuration reference.
+Of course, such a change would need a different proposal to be approved before this one.
+
+It is also worth mentioning that a "template" `KafkaRebalance` custom resource doesn't add much complexity to the operator in terms of watching, reacting and processing.
+The `KafkaRebalanceAssemblyOperator` is just going to ignore such a resource when it's created, updated or deleted.
+It can be only referenced by an "actual" `KafkaRebalance` custom resource so its creation doesn't trigger any action by the operator.
+The operator only reads the "template" `KafkaRebalance` custom resource and generates a new corresponding "actual" `KafkaRebalance` custom resource when the cluster is being scaled.
+
+### Extending the `Kafka.status` field with auto-rebalancing related state
+
+The auto-rebalancing, together with the scaling action, can take several reconciliations to finish.
+For this reason, taking into account cluster operator might crash as well, we need to save some information about the state of the auto-rebalancing operation.
+The `Kafka` custom resource is extended by adding a new `status.autoRebalance` field.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # ...
+status:
+  autoRebalance:
+    lastTransitionTime: # timestamp of last transition during the auto-rebalancing flow
+    state: # if a scaling operation is running or what's the state of the current rebalancing
+    modes:
+      - mode: add-brokers
+        brokers: # list with brokers' IDs to add (as part of the scaling up)
+      - mode: remove-brokers
+        brokers: # list with brokers' IDs to remove (as part of the scaling down)
+```
+
+The `state` field provides the current state of the rebalancing operation, reflecting if it's for a scale down (`RebalancingOnScaleDown` value) or scale up (`RebalancingOnScaleUp` value).
+When the rebalancing is running, its value is inferred from the status of the corresponding "actual" `KafkaRebalance` custom resource that was created for that purpose.
+The `lastTransitionTime` field provides the timestamp of the latest auto-rebalancing state update.
+The `modes` field lists the enabled rebalancing modes, as defined through the `spec.cruiseControl.autoRebalance.mode` field, with the corresponding `brokers` list containing the brokers' IDs that are part of the scaling (up/down) actions to be useful across reconciliations.
```

Comment:
They are the enabled modes kind of mirroring what we have in the spec. This part is mostly for storing the brokers' IDs so I need the mode to which they are related to.

Reply:
Why do you need it if there is no scaling happening? I think we should keep the status lean -> when nor autorebalance is happening, the status should IMHO not be present at all.

Reply:
I see your point but, while avoiding the mode/brokers could work for scale down because it provides you the list of removed brokers each time, it can't work for scale up. If scale up happens but the rebalancing can't happen in the same reconciliation for any reasons (there is a rebalancing for scale down already running, operator restarts, ...), we need to have the brokers list in the status for a future reconciliation (because it won't be provided anymore). So I can understand that status should have what's going on but in this case we need those info soon. Related to another comment from Kate, I am thinking I could need a `state` for each mode and not for the overall auto-rebalancing. This `state` could clarify more the overall auto-rebalancing status. I am still thinking about it.

Reply:
But most of the time, there is no scale-up in progress and it would be anyway empty. So why do you want to keep it there? The same applies for the state. What is the state when nothing is going on in the auto-rebalancing?

Reply:
I am not saying that the mode/brokers fields are there since the beginning. What I am saying is that they have to be there even if the actual rebalancing is not running yet. So the delayed scale-up example which is important to get the brokers to scale up when the list is empty.
So if the user doesn't do a scale down and up at the same time, but just one, the mode/brokers list will contain just the one running.

Reply:
You are saying `The modes field lists the enabled rebalancing modes` -> I cannot help myself but I read it only one way ... if the mode is enabled it is listed. So maybe you should rephrase it to something like `The modes field lists contains the enabled modes with some ongoing or planned auto-rebalancing ...`?

Reply:
I agree with Jakub, based on my understanding of what you describe later in the proposal what about:
```suggestion
The `modes` field is used to store a list of brokers' IDs that were part of the scaling (up/down) action that triggered an auto-rebalance to be used by the operator across reconciliations. For each enabled rebalancing mode, as defined through the `spec.cruiseControl.autoRebalance.mode` field, it contains either:
  - the brokers' IDs relevant to the current ongoing auto-rebalance, or
  - the brokers' IDs relevant to a queued auto-rebalance (if a previous auto-rebalance is still in progress)
```

Reply:
I got you Jakub. Yeah maybe it was misleading. I will use Keith suggestion, which is pretty clear and descriptive.